File: gateway/session.py
from __future__ import annotations

import asyncio
import json
import logging
import sqlite3
import uuid
from contextlib import contextmanager
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from pathlib import Path

log = logging.getLogger(__name__)

_DDL = """
CREATE TABLE IF NOT EXISTS sessions (
    id TEXT PRIMARY KEY,
    peer_id TEXT NOT NULL,
    channel TEXT NOT NULL,
    messages TEXT NOT NULL DEFAULT '[]',
    last_active TEXT NOT NULL
);
CREATE INDEX IF NOT EXISTS idx_peer_channel ON sessions(peer_id, channel);
"""


@dataclass
class Session:
    id: str
    peer_id: str
    channel: str
    messages: list = field(default_factory=list)
    last_active: datetime = field(default_factory=datetime.utcnow)
    _lock: asyncio.Lock = field(default_factory=asyncio.Lock, repr=False)


class SessionStore:
    def __init__(self, db_path: Path, scope: str = "per-peer", idle_hours: int = 8, reset_hour: int = 4):
        self._db_path = db_path
        self._scope = scope
        self._idle_hours = idle_hours
        self._reset_hour = reset_hour
        self._cache: dict[str, Session] = {}
        db_path.parent.mkdir(parents=True, exist_ok=True)
        with self._conn() as conn:
            conn.executescript(_DDL)

    @contextmanager
    def _conn(self):
        conn = sqlite3.connect(self._db_path)
        conn.row_factory = sqlite3.Row
        try:
            yield conn
            conn.commit()
        finally:
            conn.close()

    def _key(self, peer_id: str, channel: str) -> str:
        if self._scope == "per-peer":
            return peer_id
        return f"{channel}:{peer_id}"

    def get_or_create(self, peer_id: str, channel: str) -> Session:
        key = self._key(peer_id, channel)
        session = self._cache.get(key)

        if session and self._is_expired(session):
            log.info("Session expired for %s, resetting", key)
            session = None
            del self._cache[key]

        if session is None:
            session = self._load(key, peer_id, channel) or Session(
                id=str(uuid.uuid4()),
                peer_id=peer_id,
                channel=channel,
            )
            self._cache[key] = session

        return session

    def _is_expired(self, session: Session) -> bool:
        now = datetime.utcnow()
        if now - session.last_active > timedelta(hours=self._idle_hours):
            return True
        # daily reset at configured hour
        reset_today = now.replace(hour=self._reset_hour, minute=0, second=0, microsecond=0)
        if session.last_active < reset_today <= now:
            return True
        return False

    def _load(self, key: str, peer_id: str, channel: str) -> Session | None:
        with self._conn() as conn:
            row = conn.execute(
                "SELECT * FROM sessions WHERE peer_id=? AND channel=?",
                (peer_id if self._scope == "per-peer" else peer_id, channel)
            ).fetchone()
        if not row:
            return None
        s = Session(
            id=row["id"],
            peer_id=row["peer_id"],
            channel=row["channel"],
            messages=json.loads(row["messages"]),
            last_active=datetime.fromisoformat(row["last_active"]),
        )
        return s if not self._is_expired(s) else None

    def save(self, session: Session) -> None:
        session.last_active = datetime.utcnow()
        with self._conn() as conn:
            conn.execute(
                "INSERT OR REPLACE INTO sessions(id, peer_id, channel, messages, last_active) VALUES(?,?,?,?,?)",
                (session.id, session.peer_id, session.channel,
                 json.dumps(session.messages), session.last_active.isoformat()),
            )

File: src/gateway/session.ts
import { Database } from "bun:sqlite";
import { mkdirSync } from "fs";
import { resolve } from "path";
import { randomUUID } from "crypto";
import type { Message } from "../providers/base.ts";

export interface Session {
  id: string;
  peerId: string;
  channel: string;
  messages: Message[];
  lastActive: Date;
}

export class SessionStore {
  private readonly db: Database;
  private readonly cache = new Map<string, Session>();
  // tracks how many messages per session have already been flushed to DB
  private readonly flushedCount = new Map<string, number>();

  constructor(
    dbPath: string,
    private readonly scope: string = "per-peer",
    private readonly idleHours: number = 8,
    private readonly resetHour: number = 4,
  ) {
    mkdirSync(resolve(dbPath, ".."), { recursive: true });
    this.db = new Database(dbPath);
    this._migrate();
  }

  private _migrate(): void {
    // Check if old schema exists (sessions.messages column)
    const hasOldCol = this.db
      .query<{ name: string; [k: string]: unknown }, []>("PRAGMA table_info(sessions)")
      .all()
      .some((r) => r.name === "messages");

    if (hasOldCol) {
      // Migrate existing rows into the new messages table, then drop the column.
      // SQLite doesn't support DROP COLUMN before 3.35, so we recreate the table.
      this.db.exec(`
        CREATE TABLE IF NOT EXISTS sessions_new (
          id          TEXT PRIMARY KEY,
          peer_id     TEXT NOT NULL,
          channel     TEXT NOT NULL,
          last_active TEXT NOT NULL
        );
        CREATE TABLE IF NOT EXISTS messages (
          session_id TEXT NOT NULL,
          seq        INTEGER NOT NULL,
          role       TEXT NOT NULL,
          content    TEXT NOT NULL,
          PRIMARY KEY (session_id, seq),
          FOREIGN KEY (session_id) REFERENCES sessions_new(id)
        );
        CREATE INDEX IF NOT EXISTS idx_peer_channel ON sessions_new(peer_id, channel);
      `);

      // Copy old sessions, migrating their message blobs into the messages table
      const oldRows = this.db
        .query<
          { id: string; peer_id: string; channel: string; messages: string; last_active: string },
          []
        >("SELECT * FROM sessions")
        .all();

      for (const row of oldRows) {
        this.db
          .query("INSERT OR IGNORE INTO sessions_new(id, peer_id, channel, last_active) VALUES(?,?,?,?)")
          .run(row.id, row.peer_id, row.channel, row.last_active);

        const msgs = JSON.parse(row.messages) as Message[];
        for (let i = 0; i < msgs.length; i++) {
          const msg = msgs[i]!;
          this.db
            .query("INSERT OR IGNORE INTO messages(session_id, seq, role, content) VALUES(?,?,?,?)")
            .run(row.id, i, msg.role, JSON.stringify(msg.content));
        }
      }

      this.db.exec(`
        DROP TABLE sessions;
        ALTER TABLE sessions_new RENAME TO sessions;
      `);
    } else {
      // Fresh install — create tables directly
      this.db.exec(`
        CREATE TABLE IF NOT EXISTS sessions (
          id          TEXT PRIMARY KEY,
          peer_id     TEXT NOT NULL,
          channel     TEXT NOT NULL,
          last_active TEXT NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_peer_channel ON sessions(peer_id, channel);
        CREATE TABLE IF NOT EXISTS messages (
          session_id TEXT NOT NULL,
          seq        INTEGER NOT NULL,
          role       TEXT NOT NULL,
          content    TEXT NOT NULL,
          PRIMARY KEY (session_id, seq),
          FOREIGN KEY (session_id) REFERENCES sessions(id)
        );
      `);
    }
  }

  private key(peerId: string, channel: string): string {
    return this.scope === "per-peer" ? peerId : `${channel}:${peerId}`;
  }

  private isExpired(session: Session): boolean {
    const now = new Date();
    const idleMs = this.idleHours * 60 * 60 * 1000;
    if (now.getTime() - session.lastActive.getTime() > idleMs) return true;

    const resetToday = new Date(now);
    resetToday.setHours(this.resetHour, 0, 0, 0);
    if (session.lastActive < resetToday && resetToday <= now) return true;

    return false;
  }

  private loadFromDb(peerId: string, channel: string): Session | null {
    const row = this.db
      .query<{ id: string; peer_id: string; channel: string; last_active: string }, [string, string]>(
        "SELECT * FROM sessions WHERE peer_id=? AND channel=?",
      )
      .get(peerId, channel);

    if (!row) return null;

    const msgRows = this.db
      .query<{ role: string; content: string }, [string]>(
        "SELECT role, content FROM messages WHERE session_id=? ORDER BY seq",
      )
      .all(row.id);

    const messages = msgRows.map((m) => ({
      role: m.role as Message["role"],
      content: JSON.parse(m.content) as Message["content"],
    }));

    const session: Session = {
      id: row.id,
      peerId: row.peer_id,
      channel: row.channel,
      messages,
      lastActive: new Date(row.last_active),
    };

    if (this.isExpired(session)) return null;

    // seed flushedCount so save() knows these rows are already in DB
    this.flushedCount.set(session.id, messages.length);
    return session;
  }

  getOrCreate(peerId: string, channel: string): Session {
    const key = this.key(peerId, channel);
    let session = this.cache.get(key);

    if (session && this.isExpired(session)) {
      console.log(`Session expired for ${key}, resetting`);
      this.flushedCount.delete(session.id);
      session = undefined;
      this.cache.delete(key);
    }

    if (!session) {
      session = this.loadFromDb(peerId, channel) ?? {
        id: randomUUID(),
        peerId,
        channel,
        messages: [],
        lastActive: new Date(),
      };
      this.cache.set(key, session);
    }

    return session;
  }

  save(session: Session): void {
    session.lastActive = new Date();
    const already = this.flushedCount.get(session.id) ?? 0;
    const newMessages = session.messages.slice(already);
    const sessionId = session.id;
    const peerId = session.peerId;
    const channel = session.channel;
    const lastActive = session.lastActive.toISOString();
    const nextCount = session.messages.length;

    // update in-memory count immediately so concurrent calls don't double-insert
    this.flushedCount.set(sessionId, nextCount);

    // fire-and-forget — don't block the message handler
    queueMicrotask(() => {
      this.db
        .query("INSERT OR REPLACE INTO sessions(id, peer_id, channel, last_active) VALUES(?,?,?,?)")
        .run(sessionId, peerId, channel, lastActive);

      for (let i = 0; i < newMessages.length; i++) {
        const msg = newMessages[i]!;
        this.db
          .query("INSERT OR IGNORE INTO messages(session_id, seq, role, content) VALUES(?,?,?,?)")
          .run(sessionId, already + i, msg.role, JSON.stringify(msg.content));
      }
    });
  }
}
